Extension/src/LanguageServer/copilotProviders.ts

```diff
@@ -0,0 +1,136 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+'use strict';
+
+import * as vscode from 'vscode';
+import * as util from '../common';
+import * as telemetry from '../telemetry';
+import { ChatContextResult, GetIncludesResult } from './client';
+import { getActiveClient } from './extension';
+
+let isRelatedFilesApiEnabled: boolean | undefined;
+
+export interface CopilotTrait {
+    name: string;
+    value: string;
+    includeInPrompt?: boolean;
+    promptTextOverride?: string;
+}
+
+export interface CopilotApi {
+    registerRelatedFilesProvider(
+        providerId: { extensionId: string; languageId: string },
+        callback: (
+            uri: vscode.Uri,
+            context: { flags: Record<string, unknown> },
+            cancellationToken: vscode.CancellationToken
+        ) => Promise<{ entries: vscode.Uri[]; traits?: CopilotTrait[] }>
+    ): Disposable;
+}
+
+export async function registerRelatedFilesProvider(): Promise<void> {
+    if (!await getIsRelatedFilesApiEnabled()) {
+        return;
+    }
+
+    const api = await getCopilotApi();
+    if (util.extensionContext && api) {
+        try {
+            for (const languageId of ['c', 'cpp', 'cuda-cpp']) {
+                api.registerRelatedFilesProvider(
+                    { extensionId: util.extensionContext.extension.id, languageId },
+                    async (_uri: vscode.Uri, context: { flags: Record<string, unknown> }, token: vscode.CancellationToken) => {
```

Comment:
Doesn't necessarily need to be in this PR, but before we start the experiment, we'll need to wire up `flags` to this logic. 

If you want to do that now, it would simply be a matter of adding an if statement so that we only return traits if something like `context.flags['cpptraits']` is `true`. The exact name of the flag doesn't matter, and we'll make corresponding changes in Copilot to actually check with the experimentation service. We could also have multiple flags if we want to control traits individually.

Reply:
I have that implemented like this, would that work?
![image](https://github.com/user-attachments/assets/9a7906fe-3d1c-47c9-baf9-aaec90140106)

Reply:
Sorry, my bad! Yes, that looks perfect.